go-controller/pkg/ovn/host_subnet_allocator.go

```diff
@@ -0,0 +1,177 @@
+package ovn
+
+import (
+	"fmt"
+	"net"
+
+	"k8s.io/klog/v2"
+	utilnet "k8s.io/utils/net"
+
+	"github.com/ovn-org/ovn-kubernetes/go-controller/pkg/config"
+	"github.com/ovn-org/ovn-kubernetes/go-controller/pkg/metrics"
+	"github.com/ovn-org/ovn-kubernetes/go-controller/pkg/ovn/subnetallocator"
+)
+
+type hostSubnetAllocator struct {
+	// Don't inherit from BaseSubnetAllocator to ensure users of
+	// hostSubnetAllocator can't directly call the underlying methods
+	base *subnetallocator.BaseSubnetAllocator
+}
+
+func NewHostSubnetAllocator() *hostSubnetAllocator {
+	return &hostSubnetAllocator{
+		base: subnetallocator.NewSubnetAllocator(),
+	}
+}
+
+func (sna *hostSubnetAllocator) InitRanges(subnets []config.CIDRNetworkEntry) error {
+	for _, entry := range subnets {
+		if err := sna.base.AddNetworkRange(entry.CIDR, entry.HostSubnetLength); err != nil {
+			return err
+		}
+		klog.V(5).Infof("Added network range %s to host subnet allocator", entry.CIDR)
+	}
+
+	// update metrics for host subnets
+	v4count, _, v6count, _ := sna.base.Usage()
+	metrics.RecordSubnetCount(float64(v4count), float64(v6count))
+	return nil
+}
+
+// MarkSubnetsAllocated will mark the given subnets as already allocated by
+// the given owner. Marking is all-or-nothing; if marking one of the subnets
+// fails then none of them are marked as allocated.
+func (sna *hostSubnetAllocator) MarkSubnetsAllocated(nodeName string, subnets ...*net.IPNet) error {
+	if err := sna.base.MarkAllocatedNetworks(nodeName, subnets...); err != nil {
+		return err
+	}
+	_, v4used, _, v6used := sna.base.Usage()
+	metrics.RecordSubnetUsage(float64(v4used), float64(v6used))
+	return nil
+}
+
+// AllocateNodeSubnets either validates existing node subnets against the allocators
+// ranges, or allocates new subnets if the node doesn't have any yet, or returns an error
+func (sna *hostSubnetAllocator) AllocateNodeSubnets(nodeName string, existingSubnets []*net.IPNet, ipv4Mode, ipv6Mode bool) ([]*net.IPNet, []*net.IPNet, error) {
```

Comment:
suggestion for your consideration but feel free to ignore: consider an enum for ipv4, ipv6, dualstack.

Reply:
@jcaamano I think we actually want a bitfield here (which the bools emulate) or we'd need helper functions like IsIPv4Mode() etc, since in a bunch of places we do stuff like "if ipv4Mode { something } \n if ipv6Mode { something }" that handles either v4 or v6 or dual. Also we pass config.IPv4Mode etc directly into this in a couple places, so we'd have to have a conversion function from that to the enum.

Seems like if we want go enum, we should do it everywhere and not just in this PR. Thoughts?

Reply:
I see. You are right, it would not be practical to do it in this PR.